doc/decisions/1_in_discussion/internal_cache.md

```diff
@@ -0,0 +1,501 @@
+  `dest->keyUSize` is set to the same value as `source->keyUSize`.
+  `KEY_FLAG_COW_NAME` is set on `dest->flags`.
+  `KEY_FLAG_RO_NAME` is set on `source->flags`.
+  `dest->originalName` is set to `source`.
+  `source->refs` is incremented.
+
+- `dest->meta` points to a **new** keyset.
+  The keys in `dest->meta` are also copied with `KEY_CP_COW`, i.e. they are also copy-on-write keys.
+  `KEY_FLAG_COW_META` is set on `dest->flags`.
+  `KEY_FLAG_RO_META` is set on `source->flags`.
+
+The source key will remain as a read-only key.
+This constraint is needed, because the source key is the only key we can free the resources on.
+If the data or the name would change in the source key, all COW-copied keys would suddenly have another value.
+For the same reason, the source key will need to live longer than all COW-copied keys from it.
+
+A `keyCopy()` without `KEY_CP_COW` from an COW key will create a deep copy of the key.
+These keys are "normal" non-COW keys and can live on their own.
+
+Every `key*()` method that modifies data on a COW-copied key will need to allocate new memory for this data and remove the `KEY_FLAG_COW_DATA` flag.
+Every `key*()` method that modifies the name of a COW-copied key will need to allocate new memory for this name and remove the `KEY_FLAG_COW_NAME` flag.
+Every `key*()` method that modifies metadata needs to make sure that the same happens for metakeys.
+
+Keysets are not copy-on-write.
+A `ksDeepDup()` of a keyset with COW keys will create a keyset with deep-copied non-COW keys.
+Internally we may need a `ksCowDup()` function to create a keyset with copy-on-write keys from another keyset.
+Whether this function will be part of the public API is a point for discussion.
+
+### Full-blown copy-on-write implementation
+
+Make Elektra's `Key` and `KeySet` data structures copy-on-write.
+This requires some major refactoring of code within `libelektra-core`.
+Code that does only interact with the data structures via the public `libelektra-core` API should not notice any differences.
+The `mmapstorage` plugin needs to be fixed.
+
+Unlike "In-Memory COW Cache" keyDup, keyCopy, ksCopy and ksDup will always use COW.
+`ksCopy` and `ksDup` is needed for (de)duplication of metadata.
+Furthermore, the API has better usability if Key and KeySet behave the same, especially for bindings where duplication might happen implicitly.
+
+#### Changes to `Key`
+
+For the `Key`, we need to extract everything for the data and name into their own structs.
+This is done for memory-management reasons, as we need to track how many keys point to the same data and/or name.
+
+```c
+struct _KeyData {
+    union {
+        char * c;
+        void * v;
+    } data;
+
+    size_t dataSize;
+
+    uint16_t refs;
+};
+
+struct _KeyName {
+    char * key;
+    size_t keySize;
+
+    char * ukey;
+    size_t keyUSize;
+
+    uint16_t refs;
+};
+
+struct _Key {
+    struct _KeyData * keyData;
+    struct _KeyName * keyName;
```

Comment:
Some thoughts regarding moving name and data to separate structures:

1. If they are a separate entity, `mmapstorage` will need a flag once again for each of those. This is used to mark whether the data is in an mmap region or not. (or we find some bit somewhere that we can steal for this purpose)
2. Adding more indirections is probably not going to help performance. (I understand that we save memory here)

Reply:
I've added this to the discussion